### PR TITLE
[mm-141] - Uses the siteURL for all calls from the web app

### LINKED ISSUE #141: Use SiteURL for all calls from the webapp
If the server's SiteURL is configured with a subpath, the webapp's API calls do not work. The task here is to make this plugin's client to prefix its URLs with the SiteURL.

Here's an example of how the SiteURL may be computed:
https://github.com/mattermost/mattermost-plugin-jira/blob/19a9c2442817132b4eee5c77e259b80a40188a6a/webapp/src/selectors/index.js#L13-L26

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
#### Summary
Adds the site url to the calls in the frontend. Fixes #141 
#### Ticket Link
https://github.com/mattermost/mattermost-plugin-jitsi/issues/141

### CODE REVIEW

**Comment by maisnamraju** [comment]
@larkox can you take a look and see if its good to go.

File: webapp/src/client/client.ts [hunk 1]
```diff
@@ -5,30 +5,32 @@ import {id} from '../manifest';
...
         this.url = '/plugins/' + id;
-        if ((window as any).basename) {
-            this.baseUrl = (window as any).basename;
-        }
-        this.url = this.baseUrl + '/plugins/' + id;
+        this.getBaseURL();
+    }
+
+    getBaseURL = async () => {
+        this.baseUrl = this.baseUrl === '' ? Client4.getConfig().then((config) => config.ServiceSettings.SiteURL) : this.baseUrl;
```
Comment (larkox) [inline webapp/src/client/client.ts:16]: There is no need to use Client4 to ask for this information to the server. We can get the information in the redux store. You can see how the Zoom plugin gets the base URL here:
https://github.com/mattermost/mattermost-plugin-zoom/blob/master/webapp/src/index.js

**Review by larkox — CHANGES_REQUESTED** [review]
The approach should be slightly different, taking the url from the redux store, instead of asking the server.

File: webapp/src/client/client.ts (hunk 1 above)
Comment (mickmister) [reply webapp/src/client/client.ts:16]: We can make it so the `Client` class defined in this file is initialized with the redux store, so it always has access to the state containing the site URL

**Review by mickmister — CHANGES_REQUESTED** [review]

**Comment by maisnamrajusingh** [comment]
@sanjaydemansol  can you update the changes Michael asked for ?

File: webapp/src/client/client.ts [hunk 2]
```diff
@@ -24,7 +19,7 @@ export default class Client {
     }
 
     setUserStatus = async (userId: string | null, status: string) => {
-        return this.doPut(`${this.baseUrl}/api/v4/users/${userId}/status`, {user_id: userId, status});
+        return this.doPut(`${this.url}/api/v4/users/${userId}/status`, {user_id: userId, status});
```
Comment (larkox) [inline webapp/src/client/client.ts:22]: Is there any Client4 function in the redux repo that do this for you?

**Review by larkox — APPROVED** [review]
LGTM. Just a small improvement.

File: webapp/src/client/client.ts (hunk 2 above)
Comment (mickmister) [reply webapp/src/client/client.ts:22]: Right, we should definitely use the `Client4` function if it's available

Comment (sanjaydemansol) [reply webapp/src/client/client.ts:22]: nice catch @larkox , will fix it. There is one from redux selectors but it expects diff. set of parameters and has no explanation.
@mickmister Earlier I imported client4 directly, hence changed it to this.(inspired from plugin-zoom)

Comment (larkox) [reply webapp/src/client/client.ts:22]: `mattermost-redux/src/client/client4.ts` has a function `updateStatus` that could be used here. You can directly import Client4 from `mattermost-redux/src/client`. It should be properly configured.

Comment (sanjaydemansol) [reply webapp/src/client/client.ts:22]: @larkox you mean we should use `updateStatus` and delete function `setUserStatus`? won't it be out of scope for this PR?

Comment (larkox) [reply webapp/src/client/client.ts:22]: OK. Let's leave it like this for now, and we will open a new ticket.

File: webapp/src/index.tsx [hunk 3]
```diff
@@ -5,14 +5,16 @@ import * as React from 'react';
...
+import {getConfig} from 'mattermost-redux/selectors/entities/general';
 import Icon from './components/icon';
 import PostTypeJitsi from './components/post_type_jitsi';
 import I18nProvider from './components/i18n_provider';
 import RootPortal from './components/root_portal';
 import reducer from './reducers';
 import {startMeeting, loadConfig} from './actions';
 import {id as pluginId} from './manifest';
+import Client from './client';
+import {GlobalState} from 'mattermost-redux/types/store';
```
Comment (mickmister) [inline webapp/src/index.tsx:17]: Can we move this import up where the rest of the `mattermost-redux` imports are?

File: webapp/src/index.tsx [hunk 4]
```diff
@@ -5,14 +5,16 @@ import * as React from 'react';
 
 import {Channel} from 'mattermost-redux/types/channels';
 import {Post} from 'mattermost-redux/types/posts';
-
+import {getConfig} from 'mattermost-redux/selectors/entities/general';
```
Comment (mickmister) [inline webapp/src/index.tsx:8]: Can we add the blank line back in here, to separate the imports from `mattermost-redux` and local imports?

**Review by mickmister — APPROVED** [review]
Great work ~~@ sanjaydemansol~~ @maisnamrajusingh, I have a non-blocking request to reorganize the import order of external vs internal imports

File: webapp/src/index.tsx (hunk 4 above)
Comment (sanjaydemansol) [reply webapp/src/index.tsx:8]: Sure

File: webapp/src/index.tsx (hunk 3 above)
Comment (mickmister) [reply webapp/src/index.tsx:18]: @sanjaydemansol The below request is completed but this one still needs to be implemented

Comment (sanjaydemansol) [reply webapp/src/index.tsx]: @mickmister done

**Review by mickmister — APPROVED** [review]
LGTM

**Review by sibasankarnayak — APPROVED** [review]
LGTM!

**Comment by hanzei** [comment]
@dipak-demansol Could you please review this PR so we can wrap up the v2.1.0 release?

**Comment by DHaussermann** [comment]
@larkox or @maisnamraju, @dipak-demansol  and I have reviewed this change and found no regression. However, we noticed that using the plugin to launch and join a meeting on a subpath server also seems to work as expected on the `master` branch. 

Can you please confirm the issue that was solved in https://github.com/mattermost/mattermost-plugin-jitsi/issues/141 
We may need guidance or test steps on this PR to validate that the issue observed has been resolved.

**Review by dipak-demansol — CHANGES_REQUESTED** [review]
https://github.com/mattermost/mattermost-plugin-jitsi/pull/197#issuecomment-988823087

**Comment by hanzei** [comment]
@DHaussermann I sill think this PR is a code improvement. Do you have objections on merging it?

**Comment by DHaussermann** [comment]
Hi @maisnamrajusingh  and @hanzei 
We have regression tested this PR and found no issues with the change. 

Jitsi working on a subpath server can be added to release testing.  

No further functional testing is needed for this change.

LGTM!

Note that I am unable to modify label in this Repo.

**Pull request merged**